File: code/functions/function-type.c
#include <stdio.h>

// Function types in C
typedef int (*FuncType)(int, int);

int add(int a, int b) {
    return a + b;
}

int multiply(int a, int b) {
    return a * b;
}

int apply_operation(FuncType op, int x, int y) {
    return op(x, y);
}

int main() {
    FuncType operation = add;
    int result = apply_operation(operation, 5, 3);
    printf("Result: %d\n", result);
    
    operation = multiply;
    result = apply_operation(operation, 5, 3);
    printf("Result: %d\n", result);
    
    return 0;
}

File: code/functions/function-type.go
package main

import "fmt"

// Function types in Go
type FuncType func(int, int) int

func add(a, b int) int {
	return a + b
}

func multiply(a, b int) int {
	return a * b
}

func applyOperation(op FuncType, x, y int) int {
	return op(x, y)
}

func main() {
	var operation FuncType = add
	result := applyOperation(operation, 5, 3)
	fmt.Printf("Result: %d\n", result)

	operation = multiply
	result = applyOperation(operation, 5, 3)
	fmt.Printf("Result: %d\n", result)
}